oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/TokenLifecycleManager.java

```diff
@@ -0,0 +1,257 @@
+     * @param executor An optional {@link Executor} that facilitates asynchronous supply and update operations.
+     *                 A new single thread {@link Executor} will be created using
+     *                 {@link Executors#newSingleThreadExecutor()} if the {@code null}.
+     */
+    TokenLifecycleManager(BiPredicate<? super T, Instant> validator,
+                          Predicate<? super T> isUpdatable, Predicate<Throwable> shallResupply,
+                          Function<? super T, CompletionStage<? extends T>> serviceSupplier,
+                          @Nullable Function<? super T, CompletionStage<? extends T>> serviceUpdater,
+                          @Nullable Supplier<? extends T> storageSupplier,
+                          @Nullable Consumer<? super T> storageConsumer,
+                          @Nullable Executor executor) {
+        ref = new AtomicReference<>();
+        serialFuture = new SerialFuture(executor);
+        this.validator = requireNonNull(validator, "validator");
+        this.isUpdatable = requireNonNull(isUpdatable, "isUpdatable");
+        this.shallResupply = requireNonNull(shallResupply, "shallResupply");
+        this.serviceSupplier = requireNonNull(serviceSupplier, "serviceSupplier");
+        this.serviceUpdater = serviceUpdater;
+        this.storageSupplier = storageSupplier;
+        this.storageConsumer = storageConsumer;
+    }
+
+    /**
+     * Provides managed token object asynchronously. This operation may involve initial token supply or
+     * token update operation, if required.
+     */
+    public CompletionStage<T> get() {
+        final T t = ref.get();
+        if (t != null) {
+            // token already present
+            return validateOrUpdate(t);
+        }
+
+        // token is not yet present
+        // try to supply an initial token serially using a serial execution
+        return serialFuture.executeAsync(this::supplySerially);
+    }
+
+    private CompletionStage<T> supplySerially() {
+        // re-check if the token already present
+        T t = ref.get();
+        if (t != null) {
+            // token already present
+            return validateOrUpdate(t, false);
+        }
+
+        // token not yet loaded
+        // try loading from storage
+        t = load();
+        if (t != null) {
+            // token loaded
+            return validateOrUpdate(t, true);
+        }
+
+        // token has never been supplied
+        // otherwise, supply initial value
+        return supplyInitial();
+    }
+
+    /**
+     * Updates the token object if it's invalid at the given {@link Instant} time.
+     */
+    private CompletionStage<T> updateSerially(Instant instant) {
+        // after acquiring the lock, re-check if it's a valid token
+        final T t = ref.get();
+        if (validator.test(t, instant)) {
+            // simply return a valid token
+            return CompletableFuture.completedFuture(t);
```

Comment:
Global comment: Could we use `UnmodifiableFuture.completedFuture()` wherever possible, so that a caller cannot complete or cancel the returned future arbitrarily?

Reply:
I was not aware of this class. But I prefer to stick to Java API whenever possible.